parser/parser.go

```diff
@@ -1108,27 +1114,87 @@ func (p *Parser) parseFor() stmt.Stmt {
 }
 
 func (p *Parser) parseSwitch() stmt.Stmt {
+	s1, s2, isTypeSwitch := p.parseSwitchHeader()
+	if isTypeSwitch {
+		return p.parseTypeSwitch(s1, s2)
+	}
+	return p.parseExprSwitch(s1, s2)
+}
+
+func (p *Parser) parseSwitchHeader() (stmt.Stmt, stmt.Stmt, bool) {
 	p.expect(token.Switch)
 	p.next()
 
-	s := new(stmt.Switch)
+	var (
+		// consider:
+		// switch <s1>; <s2> { ... }
+		// from s1 and s2, we need to decide whether we are dealing with
+		// an expression-switch or a type-switch.
+		s1           stmt.Stmt
+		s2           stmt.Stmt
+		isTypeSwitch = false
+	)
 
 	if p.s.Token != token.LeftBrace {
 		p.noCompLit = true
-		s.Init = p.parseSimpleStmt()
-		if p.s.Token == token.Semicolon {
+		s1 = p.parseSimpleStmt()
+		switch p.s.Token {
+		case token.Semicolon:
+			// switch x := foo(); x             { ... } // 1)
```

Comment:
what are these 1), 2), 3) referencing?

Reply:
they are referenced later on in the other switch.
I'll put them instead in each of the switch-cases.

Reply:
done.